nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_local_costmap_clear_needed.hpp

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2022 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_BEHAVIOR_TREE__PLUGINS__CONDITION__IS_LOCAL_COSTMAP_CLEAR_NEEDED_HPP_
+#define NAV2_BEHAVIOR_TREE__PLUGINS__CONDITION__IS_LOCAL_COSTMAP_CLEAR_NEEDED_HPP_
+
+#include <string>
+
+#include "nav2_msgs/action/follow_path.hpp"
+#include "nav2_behavior_tree/plugins/condition/generic_error_code_condition.hpp"
+
+namespace nav2_behavior_tree
+{
+
+class IsLocalCostmapClearNeeded : public GenericErrorCodeCondition
```

Comment:
This name is pretty decent. I thought of this more of "would be helpful" versus "needed" but I suppose that's also a way of thinking about it. 

I've generally thought about the "contextual recoveries" in the BT under the task server failures to be "anything" but right now by default its only clearing the environmental models because we don't have anything else specific to do yet (which we could now with error codes). 

So question (not a request but a discussion topic) should this be targeted to the Costmap or be targeted to the Controller? E.g. from these controller codes, does that mean we should clear the costmap necessarily, or does it mean we should enact some recovery actions / primitive behaviors - which could include costmap clearing? 

If we think this can/should be used outside of just clearing the costmap context (e.g. enacting "useful" behaviors after a controller server failure which may be actionable based on error codes), should this be generalized to the controller server instead of the costmap?

Reply:
That is a good question :). 

I think for now we should have error code conditions that map to the server as you suggested. 
`wouldAControllerRecoveryHelp`
`wouldAPlannerRecoveryHelp`
.....

If we end up at a point when a error code set X could be resolved by recovery Y then we could add particular condition nodes like 
`IsLocalCostmapClearNeeded`. 

(A decorator might be bettor :)).

Reply:
OK - I think that's sensible.